crates/next-api/src/client_references.rs

```diff
@@ -24,50 +32,175 @@ pub enum ClientReferenceMapType {
     ServerComponent(ResolvedVc<NextServerComponentModule>),
 }
 
-#[turbo_tasks::value(transparent)]
-pub struct ClientReferencesSet(FxHashMap<ResolvedVc<Box<dyn Module>>, ClientReferenceMapType>);
+#[turbo_tasks::value]
+pub struct ClientReferencesSet {
+    pub client_references: FxHashMap<ResolvedVc<Box<dyn Module>>, ClientReferenceMapType>,
+    // All the server components in the graph.
+    server_components: FxIndexSet<ResolvedVc<NextServerComponentModule>>,
+    // All the server components that depend on each module
+    // This only includes mappings for modules with client references and the bitmaps reference
+    // indices into `[server_components]`
+    server_components_for_client_references:
+        FxHashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper>,
+}
+
+impl ClientReferencesSet {
+    /// Returns all the server components that depend on the given client reference
+    pub fn server_components_for_client_reference(
+        &self,
+        module: ResolvedVc<Box<dyn Module>>,
+    ) -> impl Iterator<Item = ResolvedVc<NextServerComponentModule>> {
+        let bitmap = &self
+            .server_components_for_client_references
+            .get(&module)
+            .expect("Module should be a client reference module")
```

Comment:
It might be possible to trigger this panic due to eventual consistency. Maybe use a result instead